### PR TITLE
Created Datasource and Resource for the BYOC page

### PR DESCRIPTION
**Pull request by shreyas3-vmw**

### CODE REVIEW

File: client/mds/infra-connector/service.go [hunk 1]
```diff
@@ -31,3 +39,152 @@ func (s *Service) GetRegionsWithDataPlanes(regionsQuery *DataPlaneRegionsQuery)
...
+	if err != nil {
+		return response, err
+	}
+
+	return response, nil
+}
+
+func (s *Service) GetDataPlaneById(id string) (model.ByocDataPlane, error) {
+	urlPath := fmt.Sprintf("%s/%s/%s", s.Endpoint, K8sCluster, id)
+	var response model.ByocDataPlane
```
Comment (sharmata-vmw) [inline client/mds/infra-connector/service.go:170]: same here, return type should be universal, just `DataPlane`. And method name only `GetDataPlane`

File: client/mds/infra-connector/service.go [hunk 2]
```diff
@@ -31,3 +39,152 @@ func (s *Service) GetRegionsWithDataPlanes(regionsQuery *DataPlaneRegionsQuery)
...
+	_, err := s.Api.Get(&reqUrl, nil, &response)
+	if err != nil {
+		return response, err
+	}
+	return response, nil
+}
+
+func (s *Service) GetDataPlaneByName(query *ByocDataPlaneQuery) (model.Paged[model.ByocDataPlane], error) {
+	urlPath := fmt.Sprintf("%s/%s", s.Endpoint, K8sCluster)
+	var response model.Paged[model.ByocDataPlane]
```
Comment (sharmata-vmw) [inline client/mds/infra-connector/service.go:154]: Name: `GetDataPlanes`
Param: `DataPlaneQuery`
Return type: `model.Paged[model.DataPlane]`

File: client/mds/infra-connector/service.go [hunk 3]
```diff
@@ -31,3 +39,152 @@ func (s *Service) GetRegionsWithDataPlanes(regionsQuery *DataPlaneRegionsQuery)
...
+	_, err := s.Api.Post(&urlPath, requestBody, &response)
+	if err != nil {
+		return &response, err
+	}
+
+	return &response, err
+}
+
+func (s *Service) GetDataPlane(id string) ([]model.MdsByocCloudProviderRegion, error) {
+	var response []model.MdsByocCloudProviderRegion
```
Comment (sharmata-vmw) [inline client/mds/infra-connector/service.go:141]: method: `GetDataPlaneRegions`
return type: `[]model.MdsDataPlaneRegion`

File: client/mds/infra-connector/service.go [hunk 4]
```diff
@@ -31,3 +39,152 @@ func (s *Service) GetRegionsWithDataPlanes(regionsQuery *DataPlaneRegionsQuery)
...
+
+	reqUrl := fmt.Sprintf("%s/%s/%s", s.Endpoint, K8sCluster, id)
```
Comment (sharmata-vmw) [inline client/mds/infra-connector/service.go:143]: why this API looks duplicate to GET dataPlane API ?

File: client/mds/infra-connector/service.go [hunk 5]
```diff
@@ -31,3 +39,152 @@ func (s *Service) GetRegionsWithDataPlanes(regionsQuery *DataPlaneRegionsQuery)
...
+
+	_, err := s.Api.Get(&reqUrl, nil, &response)
+	if err != nil {
+		return response, err
+	}
+	return response, nil
+}
+
+// CreateDataPlane - Submits a request to create dataplane
+func (s *Service) CreateDataPlane(requestBody *ByocDataPlaneCreateRequest) (*model.MdsPolicy, error) {
```
Comment (sharmata-vmw) [inline client/mds/infra-connector/service.go:125]: return type: `DataPlaneCreateRequest`

File: client/mds/infra-connector/service.go [hunk 6]
```diff
@@ -31,3 +39,152 @@ func (s *Service) GetRegionsWithDataPlanes(regionsQuery *DataPlaneRegionsQuery)
...
+	var response model.MdsPolicy
+	urlPath := fmt.Sprintf("%s/%s", s.Endpoint, K8sCluster)
+
+	_, err := s.Api.Post(&urlPath, requestBody, &response)
+	if err != nil {
+		return &response, err
+	}
+
+	return &response, err
+}
```
Comment (sharmata-vmw) [inline client/mds/infra-connector/service.go:138]: is this tested?

File: client/mds/infra-connector/service.go [hunk 7]
```diff
@@ -31,3 +39,152 @@ func (s *Service) GetRegionsWithDataPlanes(regionsQuery *DataPlaneRegionsQuery)
...
+	var response []model.MdsByocCloudProviderRegion
+
+	reqUrl := fmt.Sprintf("%s/%s", s.Endpoint, CloudProviders)
+
+	_, err := s.Api.Get(&reqUrl, nil, &response)
+	if err != nil {
+		return response, err
+	}
+	return response, nil
+}
```
Comment (sharmata-vmw) [inline client/mds/infra-connector/service.go:122]: i saw same return type below, is this correct?

File: client/mds/infra-connector/service.go [hunk 8]
```diff
@@ -31,3 +39,152 @@ func (s *Service) GetRegionsWithDataPlanes(regionsQuery *DataPlaneRegionsQuery)
 
 	return response, nil
 }
+
+func (s *Service) GetCloudAccounts(query *MdsCloudAccountsQuery) (model.Paged[model.MdsByocCloudAccount], error) {
+	var response model.Paged[model.MdsByocCloudAccount]
+	if query == nil {
```
Comment (sharmata-vmw) [inline client/mds/infra-connector/service.go:45]: return type: `MdsCloudAccount`

File: client/mds/infra-connector/service.go [hunk 9]
```diff
@@ -31,3 +39,152 @@ func (s *Service) GetRegionsWithDataPlanes(regionsQuery *DataPlaneRegionsQuery)
...
+
+	_, err := s.Api.Get(&reqUrl, query, &response)
+	if err != nil {
+		return response, err
+	}
+	return response, nil
+}
+
+func (s *Service) GetCertificates(query *MDSCertificateQuery) (model.Paged[model.MdsByocCertificate], error) {
+	var response model.Paged[model.MdsByocCertificate]
```
Comment (sharmata-vmw) [inline client/mds/infra-connector/service.go:63]: return type: `MdsCertificate`

File: client/mds/infra-connector/service.go [hunk 10]
```diff
@@ -31,3 +39,152 @@ func (s *Service) GetRegionsWithDataPlanes(regionsQuery *DataPlaneRegionsQuery)
...
+
+	_, err := s.Api.Get(&reqUrl, query, &response)
+	if err != nil {
+		return response, err
+	}
+	return response, nil
+}
+
+func (s *Service) GetTshirtSizes(query *MdsTshirtSizesQuery) (model.Paged[model.MdsByocTshirtSize], error) {
+	var response model.Paged[model.MdsByocTshirtSize]
```
Comment (sharmata-vmw) [inline client/mds/infra-connector/service.go:82]: return type: `MdsTshirtSize`

File: examples/data-sources/vmds_byoc/cloud_provider_regions/data-source.tf [hunk 11]
```diff
@@ -0,0 +1,2 @@
+data "vmds_cloud_provider_regions" "all" {
+}
```
Comment (sharmata-vmw) [inline examples/data-sources/vmds_byoc/cloud_provider_regions/data-source.tf:2]: i think the underlying API requires some params, check once

File: client/mds/infra-connector/service.go (hunk 6 above)
Comment (shreyas3-vmw) [reply client/mds/infra-connector/service.go:138]: Yes. Did not receive any error for return type policy . However changed it to correct response format.

File: examples/data-sources/vmds_byoc/cloud_provider_regions/data-source.tf (hunk 11 above)
Comment (shreyas3-vmw) [reply examples/data-sources/vmds_byoc/cloud_provider_regions/data-source.tf:2]: Verified . It does not require any parameter .

**Review by sharmata-vmw — APPROVED** [review]

**Pull request merged**